rust_api/src/routes.rs

```diff
@@ -0,0 +1,23 @@
+use tide::{Body, Request, Response};
+use mongodb::{bson::doc};
+
+use crate::document::User;
+use crate::state::SharedState;
+
+pub async fn find_user(req: Request<SharedState>) -> tide::Result<impl Into<Response>> {
+    let user_name = req.param("name")?;
+    let database = req.state().get_db().database("db");
+    let users = database.collection::<User>("users");
+
+    let filter = doc! { "name": user_name };
+    // TODO: handle the case of a user not existing by
+    // returning a nice error message in JSON instead of panicking
+    let user: User = users
+        .find_one(filter, None)
+        .await?
+        .expect("Could not find a user with that name");
```

Comment:
Måske noget à la `.map_err(UserNotFound)?` og så en custom error type som implementerer `Into<Response>`?

`expect()` markerer at man forventer at have en værdi. I dette tilfælde er det forventeligt, at en bruger ikke findes, og brugen af `expect()` stemmer derfor ikke overens med hvordan den normalt bliver brugt.

Alternativ løsning kunne være noget i stil med:
```rust
if let user = users.find_one(filter, None).await? {
    let mut res = Response::new(200);
    res.set_body(Body::from_json(&user)?);
} else {
    let mut res = Response::new(404);
    res.set_body(... json struct ...);
}
Ok(res)
```

Men det er også grimt. Jeg tror ikke at error handling story'en i Tide er fantastisk god, og jeg ville nok vælge et af de fire mest anvendte web-frameworks i stedet: Actix-web, Rocket, Warp eller min favorit: Axum.

Reply:
Hvis man vælger den alternative løsning kunne man måske gøre brug af "Everything is an expression" og lade de to branches hver returnere et `res` à la:
```rust
let res = if let user = ... { } else { };
Ok(res)
```

Reply:
Tak for reviewet super fedt at du ville kigge på det  og  jeg lige kunne få en bedre forståelse for ``expect()`` :pray:  

Ja ``map_err()`` med en custom error type vil nok være vejen frem. 
Er fristet til at give Axum et skud, har kigget på det meget hurtigt og det virker ret nice.

Reply:
No worries! Axum er mega nice, og har et ret aktivt økosystemet og samme overordnede mål som Tide (ergonomi over performance). Her er et nærmest komplet eksempel på hvordan jeres use case kan håndteres med Axum: https://github.com/tokio-rs/axum/blob/main/examples/error-handling-and-dependency-injection/src/main.rs#L60

Reply:
Tak for linket, det gjorde det lige en del lettere. Har skiftet over til Axum og det fungerer bare :ok_hand: